projects/pancakeswap/index.js

```diff
@@ -1,51 +1,122 @@
 
   const pairAddresses = result.trade_pairs.map(({ swap_pair_contract }) =>
     swap_pair_contract.toLowerCase()
-  )
+  );
 
-  return pairAddresses
+  return pairAddresses;
+}
+
+async function getPoolRewards(block) {
+  // Get reward token addresses
+  const rewardTokens = (await sdk.bsc.abi.multiCall({
+    abi: REWARD_TOKEN,
+    block,
+    calls: SYRUP_POOLS.slice(1).map(poolAddress => ({ target: poolAddress }))
+  })).output;
+
+  let rewardForPool = rewardTokens.reduce((acc, res) => {
+    if (res.success) {
+      const { target: poolAddress } = res.input;
+      // Here we make the (valid) assumption that no two pools have the same reward
+      acc[poolAddress] = res.output;
+    }
+
+    return acc;
+  }, {});
+
+  // Get reward token balances
+  const rewardBalances = (await sdk.bsc.abi.multiCall({
+    abi: "bep20:balanceOf",
+    block, 
+    calls: Object.entries(rewardForPool)
+      .map(([poolAddress, rewardAddress]) => ({ target: rewardAddress, params: poolAddress }))
+  })).output;
+
+  // Combine addresses with balances
+  return rewardBalances.reduce((acc, res) => {
+    const { target: rewardAddress } = res.input;
+
+    if (res.success && rewardAddress.toLowerCase() !== CAKE.toLowerCase()) {
+      acc[rewardAddress] = res.output;
+    }
+
+    return acc;
+  }, {});
+}
+
+async function getPoolDeposits(block) {
+  const balances = (await sdk.bsc.abi.multiCall({
+    target: CAKE,
+    abi: "bep20:balanceOf",
+    block,
+    calls: SYRUP_POOLS.map(address => ({ params: address }))
+  })).output;
+
+  return balances.reduce((acc, res) => {
+    if (res.success) {
+      const existingBalance = new BigNumber(acc[CAKE] ?? '0');
+      const additionalBalance = new BigNumber(res.output);
+
+      acc[CAKE] = existingBalance.plus(additionalBalance).toFixed();
+    }
+
+    return acc;
+  }, {});
+}
+
+async function getPoolBalances(block) {
+  const poolRewards = await getPoolRewards(block);
+  const poolDeposits = await getPoolDeposits(block);
+
+  return combineBalances(poolDeposits, poolRewards);
+}
+
+function combineBalances(b1, b2) {
+  const normal1 = Object.entries(b1).reduce((acc, [key, value]) => {
+    acc[key.toLowerCase()] = value;
+    return acc;
+  }, {});
+  const normal2 = Object.entries(b2).reduce((acc, [key, value]) => {
+    acc[key.toLowerCase()] = value;
+    return acc;
+  }, {});
+
+  const addressSet = new Set([...Object.keys(normal1), ...Object.keys(normal2)]);
+  const addresses = Array.from(addressSet);
+
+  return addresses.reduce((acc, address) => {
+    const balance1 = new BigNumber(normal1[address] ?? '0');
+    const balance2 = new BigNumber(normal2[address] ?? '0');
+    acc[address] = balance1.plus(balance2).toFixed();
+
+    return acc;
+  }, {});
 }
 
 async function tvl(_, block) {
-  const pairAddresses = await fetchPairAddresses()
+  const pairAddresses = await fetchPairAddresses();
+  //const pairAddresses = await sdk.bsc.swap.getPairAddresses(FACTORY, START_BLOCK, block.bsc);
```

Comment:
I still have it using the pancakeswap api because:
1. It is much faster than getting the pairs through the logs
2. It seems to crash when getting the balances for token1 of some of the pairs. I think this may have something to do with the fact that we get 9000 pairs from log method as opposed to around 700 with the pancakeswap api

Reply:
The crash is related to network error